src/_blog/libp2p-webrtc-browser-to-server.md

```diff
@@ -0,0 +1,248 @@
+Parity Technologies authored the [Rust](https://github.com/libp2p/rust-libp2p) implementation. Little Bear Labs focused on the [Go](https://github.com/libp2p/go-libp2p) and [JavaScript](https://github.com/libp2p/js-libp2p-webrtc) implementations. Protocol Labs led the specification work and provided reviews of the implementations.
+
+Without further ado, let's begin by introducing WebRTC and how it's currently used. Then we'll dive deep into the WebRTC implementation within libp2p.
+
+## WebRTC in the browser
+
+`Web Real-Time Communication`, commonly referred to as `WebRTC`, is a [set of standards](https://w3c.github.io/webrtc-pc/) that allows browsers, clients, and servers to connect to other peers to exchange audio, video, and data. In most cases, peers directly connect to other peers, improving privacy and requiring fewer hops than on a relay.
+
+While WebRTC handles audio, video, and data traffic, we're just going to focus on the data aspect because that's the API leveraged in libp2p-webrtc.
+
+WebRTC is built directly into browsers, so using the API is straightforward. Peers connect via an `RTCPeerConnection` interface. Once connected, `RTCDataChannel`s can be added to the connection to send and receive binary data.
+
+Peers use external [STUN](https://datatracker.ietf.org/doc/html/rfc3489) servers to determine their public address and any router restrictions that prohibit peer-to-peer communications. In the case of a restriction, [TURN](https://datatracker.ietf.org/doc/html/rfc8656) servers relay data between peers using a Signaling Channel.
+
+Once IP addresses are obtained, a peer sends an Offer [SDP](https://datatracker.ietf.org/doc/html/rfc4566) to the other peer. This Offer SDP details how the initiating peer can communicate (IP address, protocols, fingerprints, encryption, etc.). The other peer sends an Answer SDP to the initiating peer. Both peers now have enough information to start the DTLS handshake.
+
+The DTLS handshake is performed using fingerprints contained in the Offer and Answer SDPs. After the handshake is complete, data is sent between peers using the SCTP (Stream Control Transmission Protocol) protocol, encrypting messages with DTLS over UDP or TCP.
+
+
+## WebRTC in libp2p
+
+<!--
+participantspacing 20
+entryspacing 0.75
+
+participant Browser
+participant Server
+box over Server: Generate TLS Certificate
+box over Server: Listen on UDP Port
+box over Browser: Create RTCPeerConnection
+box over Browser: Create Server's Answer SDP
+box over Browser: Create Offer SDP
+Browser->Server:STUN Binding Request
+box over Server: Create Browser's Offer SDP
+
+Server->(1)Browser:DTLS Handshake
+Browser->(1)Server:DTLS Handshake
```

Comment:
Since WebRTC uses DTLS 1.2, the handshake takes 2 roundtrips. The diagram should probably reflect that.

Reply:
I can add that level of fidelity.  IIRC there are 3 RTs for DTLS (6 flights total), assuming no retransmissions.  I'm wondering if it would be OK to add some messaging around this w/o 6 lines (noisy) 🤔

Reply:
I added a box beneath the handshake to provide info about the 3 RTs.

Reply:
I'll do some digging.  This is a chart from the author of Pion: 
![image](https://user-images.githubusercontent.com/1536436/210901193-deeb643a-7e4a-400e-aa08-69a7b0ed02d2.png)

I'll need to see how Chrome does this.  I may need to have Wireshark observe the handshake to be certain.

Reply:
Wireshark output when observing the browser-to-server example in js-libp2p-webrtc:

![image](https://user-images.githubusercontent.com/1536436/210904091-e9a42bba-6428-4e3d-a20b-05322a66c239.png)

It looks like there is a total message count of 6 in the DTLS handshake(following STUN).

Reply:
What's the server here? I'm surprised to see a Hello Verify Request here. That's an anti-DoS measure in DTLS (similar to the Retry mechanism in QUIC, for those familiar with the QUIC spec). It adds 1 RTT to the handshake. If that's either rust-libp2p or go-libp2p, it would be worth investigating how / if we can turn this off. In terms of DoS defense, it doesn't buy us a lot here since we already create state when we receive the STUN packet, so doing a DoS defense in DTLS arguably comes to late.

Reply:
The server is a Go libp2p server.

Reply:
Relevant Pion code:
- https://github.com/pion/dtls/blob/master/flight1handler.go
- https://github.com/pion/dtls/blob/master/flight2handler.go
- https://github.com/pion/dtls/blob/master/pkg/protocol/handshake/message_hello_verify_request.go

Reply:
To summarize:

- During a DTLS handshake the server side MAY send a `HelloVerifyRequest`, thus adding an additional RTT to the DTLS handshake (2 RTT -> 3 RTT).
- This seems to be the default behavior of Pion.

The discussion on whether to tackle the optimization to only send `HelloVerifyRequest` when being DOS attacked is out of scope for this blog post. Further, it is an implementation decision, that can be rolled out by server implementations independently.

To move forward I suggest going with the conservative (current) 3 RTT (including `HelloVerifyRequest`), but adding a small note like the below:

``` diff
- box over Browser,Server: Full DTLS Handshake is 3 round trips
+ box over Browser,Server: Full DTLS Handshake is 3 round trips (2 without `HelloVerifyRequest` DOS prottection)
```